[14.0][ADD] mrp_bom_nested_bom_generator

geomer198 opened the pull request.

geomer198 committed:
[ADD] Add module.

elvise commented:
@geomer198 @isserver1 
CC @francesco-ooops 

Please check this 
Product4 has only legs
![Schermata 2022-05-27 alle 17 25 07](https://user-images.githubusercontent.com/5269577/170730654-d3726959-fa25-40c2-8238-9667a018c5e8.png)

But it's created with these:
![Schermata 2022-05-27 alle 17 25 55](https://user-images.githubusercontent.com/5269577/170730841-aa0feb39-6b29-405d-931d-06bcbf0ce333.png)

elvise commented:
@geomer198 @isserver1 
CC @francesco-ooops 
Product 5 only has only color
![Schermata 2022-05-27 alle 17 33 09](https://user-images.githubusercontent.com/5269577/170731952-2adcc05c-3155-455d-92ec-bdbc3719ed03.png)
Product 5 is created without attribute 
![Schermata 2022-05-27 alle 17 34 10](https://user-images.githubusercontent.com/5269577/170732013-40370525-93ac-4a9a-829b-8073104f7680.png)

geomer198 committed:
[FIX] Changed function _prepare_partner_attribute_ids and test.

francesco-ooops commented:
@geomer198 can you add description and usage?

geomer198 committed:
[IMP] readme added.

ivs-cetmix committed:
[IMP] description added

geomer198 committed:
[FIX] Changed function for prepare parent product attribute value for component attibutes.

geomer198 committed:
[FIX] Changed tests and module functions.

geomer198 committed:
[FIX] Check change product template nested bom and unlink old BOMs.

geomer198 committed:
[FIX] Changed function action_generate_nested_boms.

geomer198 committed:
[FIX] Changed field default value.

geomer198 committed:
[IMP] Added new tests.

geomer198 committed:
[FIX] Moved nested BOM from product.template to mrp.bom without tests.

geomer198 committed:
[FIX] demo.xml : fixed demo data
mrp_bom.py : fixed function name in variable func
mrp_nested_bom.py : changed func create_product

geomer198 committed:
[FIX] Changed tests
mrp_bom.py : changed func create_boms

geomer198 committed:
[FIX] USAGE.rst : changed description.

elvise commented:
@geomer198 please check travis

geomer198 committed:
[FIX] Test test_action_generate_nested_boms_valid dropped

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        inverse_name="bom_id",
+    )
+
+    nested_bom_count = fields.Integer(
```
Right now this field in not used. So no need for it.

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+    parent_bom_id = fields.Many2one(comodel_name="mrp.bom", string="Parent MRP BOM")
+
+    child_bom_ids = fields.One2many(
+        comodel_name="mrp.bom", inverse_name="parent_bom_id", string="Child MRP BOM"
```
string="Child BOMs"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        comodel_name="mrp.bom", inverse_name="parent_bom_id", string="Child MRP BOM"
+    )
+
+    def _compute_nested_bom_count(self) -> None:
```
Non needed

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        Unlink BOM by product.template
+        :return: bool
+        """
+        mrp_bom_ids = self.child_bom_ids
```
variable name is confusing. If we are dealing with child BOMs should be "child_bom_ids"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        mrp_bom_ids = self.child_bom_ids
+        if not len(mrp_bom_ids) > 0:
+            return False
+        used_mrp_bom_ids = (
```
Follow the rule "meaningful comes first".
[Exmaple](https://github.com/odoo/odoo/blob/10378872eddd6037f479fa8cbb3ed65d5e4b52c6/addons/sale/models/sale.py#L194)
so "mrp_bom_used_ids", "mrp_bom_unused_ids" etc

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+
+    def unlink_existing_bom(self) -> bool:
+        """
+        Unlink BOM by product.template
```
Cannot find any product.template in this function

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+
+    def _create_mrp_bom_record(self, product, lines):
+        """
+        Create mrp.bom record by product record and bom lines
```
Description is not clear

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        )
+
+    def _append_bom_line_components(self, lines, **kwargs):
+        """Update record components"""
```
Docstring is incomplete
What are **kwargs for?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        :return: bool
+        """
+        mrp_bom_ids = self.child_bom_ids
+        if not len(mrp_bom_ids) > 0:
```
why do we need len here? "if not mrp_bom_ids" will do the job

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+    def _create_mrp_bom_record(self, product, lines):
+        """
+        Create mrp.bom record by product record and bom lines
+        :param models.Model product: product.template record
```
Are you sure that this is "product.template" record? Line 91 tells us that this is product.product

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        """
+        Create mrp.bom record by product record and bom lines
+        :param models.Model product: product.template record
+        :param list lines: list [(0, 0, vals), ...]
```
This is not just a list. This is an Odoo x2many command. This must be specified explicitly

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+
+    def create_boms(self) -> None:
+        """
+        Create BOM for stage
```
What does it mean "for stage"?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+        self.changed_nested_bom = False
+        return True
+
+    def action_parent_bom(self):
```
Function name must be more explicit. erg "action_open_parent_bom"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+        required=True,
+    )
+
+    parent_id = fields.Many2one(
```
Naming is misleading. "parent_id" is a semi-reserved keyword that refers to the parent record of the same model.
So it must not be used in this case

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+        related="bom_id.product_tmpl_id",
+    )
+
+    all_attribute_ids = fields.Many2many(
```
Naming is not clear. Why "all"? Looks like here we have aggregated attributes: own and from product template.
so "attribute_aggregated_ids" looks more reasonable

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+    @api.depends("bom_id.nested_bom_ids", "parent_id.attribute_line_ids")
+    def _compute_all_attribute_ids(self) -> None:
+        """
+        Compute product attributes by parent product and component
```
Meaning is not clear. According to the code it should be something like "aggregate own attributes with parent product template ones"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+            )
+            parent_attributes = rec.parent_id.attribute_line_ids.mapped("attribute_id")
+            attributes_ids = default_attributes | parent_attributes
+            rec.all_attribute_ids = attributes_ids
```
Why not to assign it directly in the previous line?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+    _inherit = "mrp.bom"
+
+    bom_type_nested = fields.Boolean(string="Type Nested BOM", default=False)
+    changed_nested_bom = fields.Boolean(string="Changed Nested Bom", default=True)
```
Confusing name for a flag. Should be "nested_bom_updated"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "Nested BOM Generator",
+    "summary": "Nested BOM Generator",
+    "author": "Cetmix, Odoo Community Association (OCA)",
```
Ooops404, Cetmix, Odoo Community Association (OCA)

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "Nested BOM Generator",
+    "summary": "Nested BOM Generator",
```
Generate multi lever nested BOMs

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,157 @@
+class MrpBom(models.Model):
+    _inherit = "mrp.bom"
+
+    bom_type_nested = fields.Boolean(string="Type Nested BOM", default=False)
```
Field name can be more straight, string is not clear, "False" is the default value for Booleans
`nested_bom = fields.Boolean(string="Nested BOM")`

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+        """
+        Create product by parent product
+        :param int bom_id: mrp.bom record id
+        :return new product template id
```
This is confusing. What are you returning? id of the new product template? or new product_template record?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+
+        # Create product template if product_tmpl_id not set
+        if not product_tmpl_id:
+            vals.update(product_tmpl_id=self.create_product(bom_id))
```
This will crash in case of bom_id is False. Is this case covered by tests?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+
+        # Raising Exception if product qty less or equal zero
+        if product_qty <= 0.0:
+            raise models.ValidationError(_("Nested BOM: Product Qty must be than 0.0"))
```
...must be than.. what? bigger than? less than?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+            raise models.ValidationError(_("Nested BOM: Product Qty must be than 0.0"))
+        return super(MrpNestedBomLine, self).create(vals)
+
+    def _prepare_parent_attribute_ids(self) -> Set[int]:
```
this should be also fixed according to the "parent_id" comment above.

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+        """
+        for rec in self:
+            attr_ids = rec._prepare_parent_attribute_ids()
+            if not len(attr_ids) > 0:
```
Is len() really needed here?

geomer198 replied:
Resolved

ivs-cetmix reviewed (changes requested):
Have not checked tests yet. 
Anyway: "must be equal **to** zero/value etc" NOT "must equal zero/value/etc"
OR
"must be zero"

geomer198 committed:
[FIX] Changes:
* Model mrp.bom : deleted old field and function,  renamed variables, added function for field changed_nested_bom.
* Model mrp.nested.bom : changing change_nested_bom field, restruct func create_product.
* Test TestMrpBom : deleted function test_nested_bom_count, refactor test_create_boms_invalid function.

* Test TestMrpNestedBom : refactor test_create_product func.
* Changed mrp.bom view.

geomer198 committed:
[FIX] Changes:
* Model mrp.bom :
- unlink_existing_bom func : changed docstring and renaming vars.
- _create_mrp_bom_record func : changed docstring.
- _append_bim_line_components func : changed func argument
- create_boms func : changed docstring
- renamed function action_parent_bom to action_open_parent_bom.

* mrp.bom view : changed button argument ‘name’

geomer198 committed:
[FIX] Model mrp.nested.bom : Renamed field parent_id to bom_product_tmpl_id

geomer198 committed:
[FIX] Model mrp.nested.bom : renamed all_attribute_ids field to attribute_aggregated_ids.

geomer198 commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,188 @@
+        :return None
+        """
+        for rec in self:
+            default_attributes = rec.product_tmpl_id.attribute_line_ids.mapped(
```
Skipped

geomer198 committed:
[FIX] Model mrp.bom : renamed changed_nested_bom field to nested_bom_updated.

geomer198 committed:
[FIX] Changes:
* Model mrp.bom : renamed bom_type_nested field to nested_bom.
manifest.py : changed summary and author.

geomer198 committed:
[FIX] Model mrp.bom : changed funcs _create_mrp_bom_record.
refactor create_boms func.

geomer198 committed:
[FIX] Changes:
* Model mrp.nested.bom :
- changed create_product func docstring
- refactor create func
- rename _prepare_parent_attribute_ids func name to _prepare_bom_product_tmpl_attribute_ids
- changed _prepare_product_attribute func.

geomer198 committed:
[FIX] Changed tests.

geomer198 committed:
[FIX] Model mrp.bom : changed _create_mrp_bom_record func.
TestMrpBom : added test_create_mrp_bom_record_invalid test func.

geomer198 committed:
[FIX] TestMrpNestedBom : changed test_create_invalid test.

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+        inverse_name="bom_id",
+    )
+
+    parent_bom_id = fields.Many2one(comodel_name="mrp.bom", string="Parent MRP BOM")
```
Here we can use “parent_id” and “child_ids” names to benefit from Odoo built parent/child support

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+    CHOICE_FUNC = {True: "_create_mrp_bom_record", False: "_append_bom_line_components"}
+
+    @api.depends("nested_bom_ids")
+    def _compute_nested_bom_updated(self):
```
No idea of what this field and function are used for.
Never use “write” in computed fields especially in Odoo 13+ Use “update” instead

geomer198 replied:
NO! self is a recordset, so need used write()

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+            return False
+        self.nested_bom_ids._prepare_product_attribute()
+        self.create_boms()
+        self.nested_bom_updated = False
```
Saving value to the computed field directly

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+
+    @api.depends("nested_bom_ids")
+    def _compute_nested_bom_updated(self):
+        self.write({"nested_bom_updated": True})
```
nested_bom_updated will ALWAYS be “true”

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+        )
+        return True
+
+    def _append_bom_line_components(self, lines, **__):
```
What is this function used for? And why does it need the “**__”?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+        comodel_name="mrp.bom", inverse_name="parent_bom_id", string="Child BOMs"
+    )
+
+    CHOICE_FUNC = {True: "_create_mrp_bom_record", False: "_append_bom_line_components"}
```
What is the reason of doing it this way? Are there any chances of implementing new functions instead of extending existing ones?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+            bom_lines = component._prepare_bom_lines(nested_bom)
+            key = "bom_product_template_attribute_value_ids"
+            line_ptavs = [val[key] for _, _, val in bom_lines]
+            func = getattr(self, self.CHOICE_FUNC[bool(index)])
```
This is over complicated. A simple “if then else…” will do the job. And it will be much more clear and fast

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,154 @@
+        :return bool
+        """
+        self.ensure_one()
+        if not len(self.nested_bom_ids) > 0:
```
Why do we need len() here?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,196 @@
+            rec.attribute_aggregated_ids = default_attributes | parent_attributes
+
+    @api.model
+    def create_product(self, bom) -> int:
```
Returning a product.template() record will make this function much more extending friendly

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,196 @@
+        )
+
+    @api.model
+    def create(self, vals: Dict[str, Any]):
```
This is a standard Odoo function. No need to put the argument types explicitly

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,196 @@
+        product_tmpl_id = vals.get("product_tmpl_id", False)
+        product_qty = vals.get("product_qty", False)
+        bom_id = vals.get("bom_id", False)
+        bom = self.env["mrp.bom"].search([("id", "=", bom_id)], limit=1)
```
bom_id is a required field. No need to check it explicitly

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,196 @@
+                vals.update(product_tmpl_id=pt_id)
+
+        # Raising Exception if product qty less or equal zero
+        if product_qty <= 0.0:
```
You can use an SQL constraint for checking this

geomer198 replied:
Resolved

ivs-cetmix reviewed (changes requested):
Please keep your code simple and readable

geomer198 committed:
[FIX] Model mrp.bom : Remove nested_bom_ids field.

geomer198 committed:
[FIX] Model mrp.bom :
- Rename parent_bom_id field to parent_id
- Rename child_bom_ids field to child_ids

geomer198 committed:
Changes :
* Model mrp.bom : Changed action_generate_nested_boms func.
* Model mrp.nested.bom : Changed create_product func (changed docstring, returned value)
* TestMrpNestedBom : changed tests

geomer198 committed:
[FIX] Model mrp.nested.bom : Changed create func. TestMrpNestedBom : fixed tests.

geomer198 committed:
[FIX] Model mrp.bom :
- Remove _append_bom_line_components func.
- Changed create_boms func.

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_nested_bom.py:
```diff
@@ -0,0 +1,191 @@
+
+
+class MrpNestedBomLine(models.Model):
+    _name = "mrp.nested.bom"
```
Rename the model -> mrp.nested.bom.line

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+
+    nested_bom = fields.Boolean(string="Nested BOM")
+
+    nested_bom_ids = fields.One2many(
```
nested_bom_line_ids
string="Nested BOM Lines"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+    nested_bom = fields.Boolean(string="Nested BOM")
+
+    nested_bom_ids = fields.One2many(
+        comodel_name="mrp.nested.bom",
```
comodel=mrp.nested.bom.line

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+        )
+        return True
+
+    def create_boms(self) -> None:
```
Are you creating child BOMs here? In this case function name should be "create_child_boms"

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+
+    def create_boms(self) -> None:
+        """
+        Create Nested BOMs and unlink/archive old BOMs
```
What is exactly created here? Multiple nested BOM? Or single nested BOM with child BOMs?

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+
+    def action_generate_nested_boms(self) -> bool:
+        """
+        Generate MRP BOM by nested BOM
```
Statement is not clear

geomer198 replied:
Resolved

ivs-cetmix commented on mrp_bom_nested_bom_generator/models/mrp_bom.py:
```diff
@@ -0,0 +1,144 @@
+                else:
+                    self.update({"bom_line_ids": [line]})
+
+    def action_generate_nested_boms(self) -> bool:
```
Which records are generated here? mrp.bom or mrp.nested.bom.line?

geomer198 replied:
mrp.bom

geomer198 replied:
Renamed to action_generate_mrp_boms

geomer198 replied:
Resolved

ivs-cetmix reviewed (changes requested):
1. Fix model naming. "mrp.nested bom" is meaningless. Use "mrp.nested.bom.line" instead.
2. Check function names once again. Naming must explicitly reflect what does the function do.

geomer198 committed:
[FIX] Renamed mrp.nested.bom model to mrp.nested.bom.line

geomer198 committed:
[FIX] Model mrp.bom : Renamed nested_bom_ids field to nested_bom_ids.

geomer198 committed:
[FIX] Model mrp.bom : Renamed create_boms, action_generate_nested_boms funcs to create_child_boms and action_generate_mrp_boms.

francesco-ooops commented:
@geomer198 can you please rebase?
